### PR TITLE
IS-450 Reload rotation timestamp on each isExitTime request

### PR DESCRIPTION
**Pull request by badrogger**
### Problem 
The solution for https://github.com/skalenetwork/internal-support/issues/450 making skale-admin will write it to corresponding file directly instead of sending rpc call. The issue is that InstanceMonitor.cpp loads this file once during startup, which effectively says that timestamp won't be passed to skaled at all. 

### Changes
`isExitTime` will check the content of rotation.txt file and compare it to current block timestamp instead on relaying on `m_finishTimestamp` field.

### Performance 
Since `rotation.txt` file is only 4 bytes and reading is happening in background performance should not be affected. 
### Unit tests
Modified tests in `InstanceMonitorTest.cpp`.

### Testing 
Tested by running rotation procedure on local network.

### CODE REVIEW

File: libethereum/InstanceMonitor.cpp [hunk 1]
```diff
@@ -46,23 +46,20 @@ void InstanceMonitor::initRotationParams( uint64_t _finishTimestamp ) {
...
-    m_finishTimestamp = _finishTimestamp;
-    LOG( m_logger ) << "Set rotation time to " << m_finishTimestamp;
+    LOG( m_logger ) << "Set rotation time to " << _finishTimestamp;
 }
 
 bool InstanceMonitor::isTimeToRotate( uint64_t _finishTimestamp ) {
     if ( !fs::exists( m_rotationInfoFilePath ) ) {
         return false;
     }
-    return m_finishTimestamp <= _finishTimestamp;
```
Comment (DmytroNazarenko) [inline libethereum/InstanceMonitor.cpp:57]: It's better to write information about rotation into logs

File: libethereum/InstanceMonitor.cpp [hunk 2]
```diff
@@ -46,23 +46,20 @@ void InstanceMonitor::initRotationParams( uint64_t _finishTimestamp ) {
     std::ofstream rotationInfoFile( m_rotationInfoFilePath.string() );
     rotationInfoFile << rotationJson;
 
-    m_finishTimestamp = _finishTimestamp;
-    LOG( m_logger ) << "Set rotation time to " << m_finishTimestamp;
+    LOG( m_logger ) << "Set rotation time to " << _finishTimestamp;
 }
 
 bool InstanceMonitor::isTimeToRotate( uint64_t _finishTimestamp ) {
```
Comment (DmytroNazarenko) [inline libethereum/InstanceMonitor.cpp]: Better to rename the parameter to `currentTimestamp`/`blockTimestamp` or sth like this

**Review by DmytroNazarenko — CHANGES_REQUESTED** [review]

File: libethereum/InstanceMonitor.cpp [hunk 3]
```diff
@@ -46,23 +46,22 @@ void InstanceMonitor::initRotationParams( uint64_t _finishTimestamp ) {
...
-void InstanceMonitor::restoreRotationParams() {
-    if ( fs::exists( m_rotationInfoFilePath ) ) {
-        std::ifstream rotationInfoFile( m_rotationInfoFilePath.string() );
-        auto rotationJson = nlohmann::json::parse( rotationInfoFile );
-        m_finishTimestamp = rotationJson["timestamp"].get< uint64_t >();
-    }
+uint64_t InstanceMonitor::rotationTimestamp() const {
+    std::ifstream rotationInfoFile( m_rotationInfoFilePath.string() );
+    auto rotationJson = nlohmann::json::parse( rotationInfoFile );
+    auto timestamp = rotationJson["timestamp"].get< uint64_t >();
```
Comment (kladkogex) [inline libethereum/InstanceMonitor.cpp]: Add try catch to handle the case of corrupt file

**Review by kladkogex — COMMENTED** [review]
Change .txt to .json

File: libethereum/InstanceMonitor.cpp [hunk 4]
```diff
@@ -46,23 +46,22 @@ void InstanceMonitor::initRotationParams( uint64_t _finishTimestamp ) {
     std::ofstream rotationInfoFile( m_rotationInfoFilePath.string() );
     rotationInfoFile << rotationJson;
 
-    m_finishTimestamp = _finishTimestamp;
-    LOG( m_logger ) << "Set rotation time to " << m_finishTimestamp;
+    LOG( m_logger ) << "Set rotation time to " << _finishTimestamp;
```
Comment (kladkogex) [inline libethereum/InstanceMonitor.cpp]: Add try catch

File: libethereum/InstanceMonitor.h [hunk 5]
```diff
@@ -61,5 +56,5 @@ class InstanceMonitor {
     void reportExitTimeReached( bool _reached );
 
 private:
-    dev::Logger m_logger{ createLogger( dev::VerbosityInfo, "instance-monitor" ) };
+    mutable dev::Logger m_logger{ createLogger( dev::VerbosityInfo, "instance-monitor" ) };
```
Comment (kladkogex) [inline libethereum/InstanceMonitor.h]: mutable is not needed here

**Review by dimalit — DISMISSED** [review]

**Review by kladkogex — DISMISSED** [review]

**Review by dimalit — DISMISSED** [review]

File: libethereum/InstanceMonitor.h [hunk 6]
```diff
@@ -36,30 +36,40 @@ class InstanceMonitor {
...
+            what_str = "File " + path.string() + " is malformed or missing";
+        }
+        virtual const char* what() const noexcept override { return what_str.c_str(); }
+    };
+
+
 private:
-    dev::Logger m_logger{ createLogger( dev::VerbosityInfo, "instance-monitor" ) };
+    mutable dev::Logger m_info_logger{ createLogger( dev::VerbosityInfo, "instance-monitor" ) };
+    mutable dev::Logger m_error_logger{ createLogger( dev::VerbosityError, "instance-monitor" ) };
```
Comment (DmytroNazarenko) [inline libethereum/InstanceMonitor.h]: Change to camelCase

**Review by DmytroNazarenko — CHANGES_REQUESTED** [review]

**Review by DmytroNazarenko — APPROVED** [review]

**Review by olehnikolaiev — APPROVED** [review]

**Pull request merged**